--- a/mac80211.c
+++ b/mac80211.c
@@ -64,7 +64,7 @@ static int rtw_ops_start(struct ieee8021
 	return ret;
 }
 
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 11, 0))
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 11, 0)) && !defined(OPENWRT)
 static void rtw_ops_stop(struct ieee80211_hw *hw)
 #else
 static void rtw_ops_stop(struct ieee80211_hw *hw, bool suspend)
@@ -1028,7 +1028,7 @@ static void rtw_ops_sta_rc_update(struct
 }
 
 const struct ieee80211_ops rtw_ops = {
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 9, 0) || defined(OPENWRT)
 	.add_chanctx = ieee80211_emulate_add_chanctx,
 	.remove_chanctx = ieee80211_emulate_remove_chanctx,
 	.change_chanctx = ieee80211_emulate_change_chanctx,
--- a/main.h
+++ b/main.h
@@ -48,7 +48,9 @@ static inline void *devm_kmemdup_array(s
 
 #define OPENWRT
 
-#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0))
+#define OPENWRT
+
+#if (LINUX_VERSION_CODE < KERNEL_VERSION(6, 6, 0)) && !defined(OPENWRT)
 /**
  * abs_diff - return absolute value of the difference between the arguments
  * @a: the first argument
@@ -69,7 +71,7 @@ static inline void *devm_kmemdup_array(s
 })
 #endif
 
-#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 8, 0)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 8, 0) && !defined(OPENWRT)
 /**
  * cfg80211_ssid_eq - compare two SSIDs
  * @a: first SSID
@@ -2435,7 +2437,7 @@ bool rtw_core_check_sta_active(struct rt
 void rtw_core_enable_beacon(struct rtw_dev *rtwdev, bool enable);
 void rtw_set_ampdu_factor(struct rtw_dev *rtwdev, struct ieee80211_vif *vif,
 			  struct ieee80211_bss_conf *bss_conf);
-#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 8, 0)
+#if LINUX_VERSION_CODE < KERNEL_VERSION(6, 8, 0) && !defined(OPENWRT)
 static inline bool rtw_ssid_equal(struct cfg80211_ssid *a,
                                  struct cfg80211_ssid *b)
 {
--- a/pci.c
+++ b/pci.c
@@ -1641,7 +1641,7 @@ static const struct rtw_hci_ops rtw_pci_
 
 static int rtw_pci_request_irq(struct rtw_dev *rtwdev, struct pci_dev *pdev)
 {
-#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 10, 0))
+#if (LINUX_VERSION_CODE >= KERNEL_VERSION(6, 10, 0)) || defined(OPENWRT)
 	unsigned int flags = PCI_IRQ_INTX;
 #else
 	unsigned int flags = PCI_IRQ_LEGACY;
@@ -1720,7 +1720,7 @@ static int rtw_pci_napi_init(struct rtw_
 {
 	struct rtw_pci *rtwpci = (struct rtw_pci *)rtwdev->priv;
 
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0) || defined(OPENWRT)
 	rtwpci->netdev = alloc_netdev_dummy(0);
 	if (!rtwpci->netdev)
 		return -ENOMEM;
@@ -1747,7 +1747,7 @@ static void rtw_pci_napi_deinit(struct r
 
 	rtw_pci_napi_stop(rtwdev);
 	netif_napi_del(&rtwpci->napi);
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0) || defined(OPENWRT)
 	free_netdev(rtwpci->netdev);
 #endif
 }
--- a/pci.h
+++ b/pci.h
@@ -244,7 +244,7 @@ struct rtw_pci {
 	bool running;
 
 	/* napi structure */
-#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0)
+#if LINUX_VERSION_CODE >= KERNEL_VERSION(6, 11, 0) || defined(OPENWRT)
 	struct net_device *netdev;
 #else
 	struct net_device netdev;
